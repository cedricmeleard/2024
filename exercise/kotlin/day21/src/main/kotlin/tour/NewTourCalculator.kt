package tour

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import java.time.Duration
import java.time.LocalTime

data class Step(val time: LocalTime, val label: String, val deliveryTime: Int)

class NewTourCalculator(private var steps: List<Step>) {
    private var calculated: Boolean = false
    private var totalDeliveryTime: Double = 0.0
    private val sortedSteps = steps.sortedBy { it.time }

    fun calculate(): Either<String, String> {
        if (steps.isEmpty()) {
            return "No locations !!!".left()
        }

        val result = StringBuilder()

        if (!calculated) {

            sortedSteps.forEach { step ->
                this.totalDeliveryTime += step.deliveryTime
                result.appendLine(fLine(step))
            }

        }

        val str: String = formatDurationToHHMMSS(totalDeliveryTime)
        result.appendLine("Delivery time | $str")
        calculated = true

        return result.toString().right()
    }

    private fun formatDurationToHHMMSS(deliveryTime: Double): String {
        val duration = Duration.ofSeconds(
            deliveryTime.toLong()
        )
        return "${duration.toHours()}%02d:${duration.toMinutesPart()}%02d:${duration.toSecondsPart()}%02d"
    }


    private fun fLine(step: Step?): String {
        if (step == null) {
            throw IllegalStateException()
        }

        return step.let {
            "${it.time} : ${it.label} | ${it.deliveryTime} sec"
        }
    }

}
