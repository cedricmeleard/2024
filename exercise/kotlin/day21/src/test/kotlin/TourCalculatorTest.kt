import arrow.core.Either
import arrow.core.left
import org.approvaltests.Approvals
import tour.NewTourCalculator
import tour.Step
import tour.TourCalculator
import java.time.Duration
import java.time.LocalTime
import java.time.temporal.ChronoUnit
import kotlin.test.Test
import kotlin.test.assertEquals

class TourCalculatorTest {

    @Test
    fun calculateGoldenMasterTest() {
        var refactored: Either<String, String> = "No calculations performed".left()

        val steps = mutableListOf<Step>()
        for (i in 0 until 30) {
            steps.add(Step(
                LocalTime.of(8, 30),
                "Location ${i}",
                DeliveryDuration(i * 26, ChronoUnit.MINUTES)))

            val origin = TourCalculator(steps).calculate();
            refactored = NewTourCalculator(steps).calculate();

            assertEquals(origin,refactored)
        }

        Approvals.verify(refactored)
    }

    private fun DeliveryDuration(duration: Int, unit: ChronoUnit)
        = Duration.of(duration.toLong(), unit).toSeconds().toInt()
}