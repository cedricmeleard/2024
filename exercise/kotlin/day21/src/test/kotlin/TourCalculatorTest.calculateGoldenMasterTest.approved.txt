Either.Right(08:30 : Location 0 | 0 sec
08:30 : Location 1 | 1560 sec
08:30 : Location 2 | 3120 sec
08:30 : Location 3 | 4680 sec
08:30 : Location 4 | 6240 sec
08:30 : Location 5 | 7800 sec
08:30 : Location 6 | 9360 sec
08:30 : Location 7 | 10920 sec
08:30 : Location 8 | 12480 sec
08:30 : Location 9 | 14040 sec
08:30 : Location 10 | 15600 sec
08:30 : Location 11 | 17160 sec
08:30 : Location 12 | 18720 sec
08:30 : Location 13 | 20280 sec
08:30 : Location 14 | 21840 sec
08:30 : Location 15 | 23400 sec
08:30 : Location 16 | 24960 sec
08:30 : Location 17 | 26520 sec
08:30 : Location 18 | 28080 sec
08:30 : Location 19 | 29640 sec
08:30 : Location 20 | 31200 sec
08:30 : Location 21 | 32760 sec
08:30 : Location 22 | 34320 sec
08:30 : Location 23 | 35880 sec
08:30 : Location 24 | 37440 sec
08:30 : Location 25 | 39000 sec
08:30 : Location 26 | 40560 sec
08:30 : Location 27 | 42120 sec
08:30 : Location 28 | 43680 sec
08:30 : Location 29 | 45240 sec
Delivery time | 188%02d:30%02d:0%02d
)